File: python/day_08.py
from pathlib import Path
import itertools
from dataclasses import dataclass


def get_data():
    return (int(x) for x in Path('data/day_08.txt').read_text().strip().split(' '))


@dataclass(frozen = True)
class Node:
    children: list
    metadata: list

    @property
    def num_children(self):
        return len(self.children)

    @property
    def num_metadata(self):
        return len(self.metadata)


def take(n, iterator):
    for _ in range(n):
        yield next(iterator)


def walk(node):
    yield node
    for child in node.children:
        yield from walk(child)


def build_node(data):
    num_children, num_metadata = take(2, data)

    return Node(
        children = list(build_node(data) for _ in range(num_children)),
        metadata = list(take(num_metadata, data)),
    )


def part_1():
    data = get_data()
    root = build_node(data)

    s = 0
    for node in walk(root):
        s += sum(node.metadata)

    return s


def value(node):
    if node.num_children == 0:
        return sum(node.metadata)

    v = 0
    for entry in node.metadata:
        try:
            v += value(node.children[entry - 1])
        except IndexError:
            pass

    return v


def part_2():
    data = get_data()
    root = build_node(data)

    return value(root)


if __name__ == '__main__':
    print(f'Part 1: {part_1()}')
    print(f'Part 2: {part_2()}')

File: ruby/day_08.rb
# frozen_string_literal: true

require 'pathname'

def read_numbers
  (Pathname(__dir__).parent / 'data' / 'day_08.txt').read.split(' ').map(&:to_i)
end

class Node
  attr_accessor :children, :metadata

  def initialize
    @children = []
    @metadata = []
  end

  def walk(&block)
    @children.each do |child|
      child.walk(&block)
    end
    yield self
  end

  def reduce(initial = 0, &block)
    acc = initial
    @children.each do |child|
      acc = child.reduce(acc, &block)
    end
    yield acc, self
  end

  def value
    return @metadata.sum if @children.length.zero?

    indices = ((@metadata.select { |m| m.between?(1, @children.length) }).map { |m| m - 1 })
    indices.reduce(0) do |acc, m|
      acc + @children[m].value
    end
  end
end

def parse(numbers, parent = nil)
  return if numbers.length.zero?

  node = Node.new
  parent.children << node unless parent.nil?

  num_children, num_metadata = numbers.shift(2)
  num_children.times do
    parse(numbers, node)
  end

  node.metadata += numbers.shift(num_metadata)
  node
end

def part_one(root)
  root.reduce do |acc, node|
    acc + node.metadata.sum
  end
end

def part_two(root)
  root.value
end

if $PROGRAM_NAME == __FILE__
  puts 'https://adventofcode.com/2018/day/8'

  numbers = read_numbers

  root = parse(numbers)

  puts "Part One: #{part_one(root)}"
  puts "Part Two: #{part_two(root)}"
end
